### PR TITLE
Work around issues with cccl main

### PR DESCRIPTION
**Pull request by miscco**
This gets cuDF build cccl main on 12.3.

There is one issue with the cuco tuple helpers but that will be fixed on the cuco side

### CODE REVIEW

File: cpp/src/io/parquet/page_string_decode.cu [hunk 1]
```diff
@@ -1186,13 +1186,13 @@ void ComputePageStringSizes(cudf::detail::hostdevice_span<PageInfo> pages,
 
   // check for needed temp space for DELTA_BYTE_ARRAY
   auto const need_sizes = thrust::any_of(
-    rmm::exec_policy(stream), pages.device_begin(), pages.device_end(), [] __device__(auto& page) {
+    rmm::exec_policy(stream), pages.device_begin(), pages.device_end(), cuda::proclaim_return_type<bool>([] __device__(auto& page) {
```
Comment (bdice) [inline cpp/src/io/parquet/page_string_decode.cu:1189]: While we're here, we can take the parameter by `const&`:
```suggestion
    rmm::exec_policy(stream), pages.device_begin(), pages.device_end(), cuda::proclaim_return_type<bool>([] __device__(auto const& page) {
```

**Review by bdice — APPROVED** [review]

**Comment by bdice** [comment]
@miscco I went ahead and applied my two (small) suggestions, so we can get CI run and merge this sooner. These aren't controversial changes so we should be able to merge soon.

**Comment by trxcllnt** [comment]
@miscco I'm using this branch in https://github.com/NVIDIA/cccl/pull/1667, will be updating this branch until that PR can build cudf.

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -53,6 +53,10 @@ repos:
       - id: clang-format
         types_or: [c, c++, cuda]
         args: ["-fallback-style=none", "-style=file", "-i"]
+        exclude: |
```
[the remark this answers is not in the document]
Comment (bdice) [reply .pre-commit-config.yaml:56]: This directory is in .gitignore. How did you trigger pre-commit on an ignored directory?

[the remark this answers is not in the document]
Comment (trxcllnt) [reply .pre-commit-config.yaml:56]: Dunno. All I know is I ran pre-commit, it said `clang-format` modified some files, but then `git status` showed nothing. Added this, did a clean build, re-ran pre-commit, and `clang-format` modified nothing. :shrug:

File: .github/workflows/pr.yaml [hunk 3]
```diff
@@ -157,8 +157,24 @@ jobs:
...
+          | tee ~/rapids-cmake-override-versions.json;
         sccache -z;
-        build-all -DBUILD_BENCHMARKS=ON --verbose;
+        build-all \
+          -j$(nproc --ignore=1) -v \
+          -DBUILD_TESTS=ON \
+          -DBUILD_BENCHMARKS=ON \
+          -DCMAKE_CXX_FLAGS="-ftemplate-backtrace-limit=0" \
+          -DCMAKE_CUDA_FLAGS="-ftemplate-backtrace-limit=0" \
+          -DRAPIDS_CMAKE_CPM_DEFAULT_VERSION_FILE="${HOME}/rapids-cmake-override-versions.json";
```
Comment (trxcllnt) [inline .github/workflows/pr.yaml:177]: Need to remove all this before merging

File: cpp/src/groupby/sort/group_rank_scan.cu [hunk 4]
```diff
@@ -134,10 +134,10 @@ std::unique_ptr<column> rank_generator(column_view const& grouped_values,
 
   auto [group_labels_begin, mutable_rank_begin] = [&]() {
     if constexpr (forward) {
-      return thrust::pair{group_labels.begin(), mutable_ranks.begin<size_type>()};
+      return thrust::make_pair(group_labels.begin(), mutable_ranks.begin<size_type>());
```
Comment (miscco) [inline cpp/src/groupby/sort/group_rank_scan.cu:137]: Those changes should not be necessary anymore

File: cpp/src/transform/row_bit_count.cu [hunk 5]
```diff
@@ -425,7 +425,7 @@ CUDF_KERNEL void compute_segment_sizes(device_span<column_device_view const> col
   // current row span - always starts at spanning over `segment_length` rows.
   auto const num_rows             = cols[0].size();
   auto const get_default_row_span = [=] {
-    return row_span{tid * segment_length, cuda::std::min((tid + 1) * segment_length, num_rows)};
```
Comment (miscco) [inline cpp/src/transform/row_bit_count.cu:428]: That should work now

File: cpp/src/join/distinct_hash_join.cu [hunk 6]
```diff
@@ -181,8 +181,12 @@ distinct_hash_join<HasNested>::inner_join(rmm::cuda_stream_view stream,
...
     thrust::make_transform_output_iterator(probe_indices->begin(), output_fn{});
 
-  auto const [probe_indices_end, _] = this->_hash_table.retrieve(
-    iter, iter + probe_table_num_rows, probe_indices_begin, build_indices_begin, stream.value());
+  auto const [probe_indices_end, _] =
+    this->_hash_table.retrieve(iter,
+                               iter + probe_table_num_rows,
+                               probe_indices_begin,
+                               build_indices_begin,
+                               cuco::cuda_stream_ref{stream.value()});
```
Comment (PointKernel) [inline cpp/src/join/distinct_hash_join.cu:189]: question: is the explicit `cuco::cuda_stream_ref` construction necessary?

Comment (PointKernel) [inline cpp/src/join/distinct_hash_join.cu:189]:
```suggestion
                               {stream.value()});
```
this should work and can avoid breaking changes once we replace the cuco stream ref to use libcudacxx's stream ref.

Comment (trxcllnt) [reply cpp/src/join/distinct_hash_join.cu:189]: Yes, otherwise the compiler thinks we're trying to call the other overload of this function.

Comment (PointKernel) [reply cpp/src/join/distinct_hash_join.cu:189]: Thanks. Can you please check my other review suggestion? That can avoid a breaking change with the upcoming cuco refactoring. The PR should be ready to go once @miscco's and my reviews are addressed.

File: cpp/src/io/orc/reader_impl_decode.cu [hunk 7]
```diff
@@ -692,8 +692,7 @@ std::vector<range> find_table_splits(table_view const& input,
      d_sizes  = d_segmented_sizes->view().begin<size_type>()] __device__(auto const segment_idx) {
       // Since the number of rows may not divisible by segment_length,
       // the last segment may be shorter than the others.
-      auto const current_length =
-        cuda::std::min(segment_length, num_rows - segment_length * segment_idx);
+      auto const current_length = std::min(segment_length, num_rows - segment_length * segment_idx);
```
Comment (trxcllnt) [inline cpp/src/io/orc/reader_impl_decode.cu:695]:
```suggestion
      auto const current_length = cuda::std::min(segment_length, num_rows - segment_length * segment_idx);
```

File: cpp/src/io/parquet/page_string_decode.cu [hunk 8]
```diff
@@ -1197,14 +1197,17 @@ void ComputePageStringSizes(cudf::detail::hostdevice_span<PageInfo> pages,
...
-  auto const need_sizes = thrust::any_of(
-    rmm::exec_policy(stream), pages.device_begin(), pages.device_end(), [] __device__(auto& page) {
-      return page.temp_string_size != 0;
-    });
+  auto const need_sizes =
+    thrust::any_of(rmm::exec_policy(stream),
+                   pages.device_begin(),
+                   pages.device_end(),
+                   cuda::proclaim_return_type<bool>(
+                     [] __device__(auto const& page) { return page.temp_string_size != 0; }));
```
Comment (trxcllnt) [inline cpp/src/io/parquet/page_string_decode.cu:1205]:
```suggestion
                     [] __device__(auto& page) { return page.temp_string_size != 0; }));
```

File: rapids_config.cmake [hunk 9]
```diff
@@ -26,6 +26,9 @@ else()
   )
 endif()
 
+set(rapids-cmake-repo trxcllnt/rapids-cmake)
+set(rapids-cmake-branch fea/cccl-2.5)
+
```
Comment (trxcllnt) [inline rapids_config.cmake:31]: Need to revert this before merging:

```suggestion
```

**Review by davidwendt — APPROVED** [review]

**Review by raydouglass — APPROVED** [review]

**Pull request merged**